### ISSUE TITLE
Default shell

### ISSUE THREAD

**Issue by mightyiam** [nix-community/home-manager#2209]
### Description

Does it make sense to set a default user shell that will be used by default in all "shell" options?

For example, in all terminal emulators and shell multiplexers.

`home.shell`, perhaps?

**Comment by rycee** [comment]
I guess it would be possible to add an option `home.shell` or something that modules like tmux could look at. But I suspect such an option may cause confusion with the user's login shell, which needs to be set on the system level (or though something like chsh).

**Comment by mightyiam** [comment]
I am currently wondering about exactly that — whether a user would be able to set their own shell somehow without need for a system level change. I noticed that the default value for [`users.users.<name>.shell`](https://search.nixos.org/options?channel=unstable&show=users.users.%3Cname%3E.shell&from=0&size=50&sort=relevance&query=users.users.%3Cname%3E.shell) is `"pkgs.shadow"`. I didn't get a good at the [`shadow` package](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/os-specific/linux/shadow/default.nix#L87) or how all this works.

One direction I'd like to explore is whether this default shell has a feature of picking up from the user's home what the user's preferred shell is. Ideally, with a system provided fallback.

Otherwise, how simply can a system be configured to behave in such a way.

In the context of NixOS, [mutable users](https://search.nixos.org/options?channel=unstable&show=users.mutableUsers&from=0&size=50&sort=relevance&query=mutable) should be considered. I am assuming that in this case, users are not able to set their own shell such as they can on common systems with `chsh`.

**Comment by rycee** [comment]
The user's shell can be set using a `shellPackage` which is a regular package derivation that has a `shellPath` attribute, in the `pkgs.shadow` case it is simply set to the [path of the `nologin` program](https://github.com/NixOS/nixpkgs/blob/439b1605227b8adb1357b55ce8529d541abbe9eb/pkgs/os-specific/linux/shadow/default.nix#L93).

In any case, I imagine it may be bad to allow any user to set the default login shell for themselves since it would be a considerable security risk. But I can imagine assigning a special "shell" to selected users, this shell would simply exec the shell configured in the user's profile, otherwise use the system runtime shell.

I.e., the user would have something like `/run/current-system/sw/bin/usershell` in their `passwd` entry. The `usershell` tool would then be responsible for actually starting up the right shell.

But I suspect there will be security considerations with that type of setup as well.

**Comment by quazar-omega** [comment]
Hello! Wanted to ask, is this left closed just because it was detected as stale or because it's actually not possible to fix?

**Comment by terlar** [comment]
There is an implementation that does basically what @rycee described:
https://github.com/viperML/noshell

**Comment by quazar-omega** [comment]
That's interesting, though I'm not sure how to make use of it when not on NixOS. It says under [Why](https://github.com/viperML/noshell?tab=readme-ov-file#why):
> It can be useful too outside of NixOS

However just before in [Usage](https://github.com/viperML/noshell?tab=readme-ov-file#usage) it says:

> Add this flake's `nixosModules.default` to your NixOS imports

So, in the end, does it enable to set your shell without modifying the system (which for me isn't NixOS) or does it need to make the first system-wide modification to then be able to expose the setting to the user?

**Comment by mightyiam** [comment]
I can't imagine it not being a system feature, first.

**Comment by quazar-omega** [comment]
I guess so ;/

**Comment by terlar** [comment]
Yeah, that is correct. It can be used on other distributions as well. But then the noshell executable has to be allowed on the system level. 

When it is in place the delegation can happen on the user side (via noshell)

The NixOS module is just a convenience for NixOS to do that plumbing. On other distributions you have to do the automation with some other tool or manually.

**Comment by quazar-omega** [comment]
Thanks for giving more details!

> The NixOS module is just a convenience for NixOS to do that plumbing. On other distributions you have to do the automation with some other tool or manually.

I see, so I'm wondering, has anyone worked out how to set it up outside NixOS then?

**Comment by terlar** [comment]
I havent tried it myself. But it should work like this:

1. Ensure the noshell is available somewhere, e.g. put it in `/usr/bin/noshell`
2. Add it to `/etc/shells`
3. Change the user shell to `/usr/bin/noshell`

Now you can control your user shell via symlinking your desired shell to `$XDG_CONFIG_HOME/shell`

**Comment by quazar-omega** [comment]
Ok I tried it now and was able to do it with Nushell, thanks!

My steps, with commands:
1. Install `noshell`
   ```
   nix profile install github:viperML/noshell
   ```
2. add it to the shells:
   ```
   sudo cp /etc/shells /etc/shells.backup
   echo $(command -v noshell) | sudo tee --append /etc/shells
   ```
3. change the user's login shell
   ```
   sudo chsh --shell $(command -v noshell) $(whoami)
   ```
4. Create a symlink to the wanted shell:
   ```
   ln -sT $(command -v nu) .config/shell # not using `$XDG_CONFIG_HOME/shell` because it is not set in my distro
   ```
5. Reboot

**Comment by terlar** [comment]
I guess the whole point was to do the last part with Home Manager, that can be achieved like this:
```
xdg.configFile."shell".source = lib.getExe pkgs.nushell;
```

**Comment by quazar-omega** [comment]
You're right, and omg I just realized something, because I *was* looking to do that, but I wasn't sure exactly how when I saw that example in the README, by skimming I mistakenly read that `pkgs.nushell` as `pkgs.noshell` so I thought that it was referencing _no_shell, but I wondered how, since it's not available in nixpkgs yet, and confused I gave up on that.  
I feel so silly now, I just had to re-read it again.

While we're at it, could I even install noshell's flake with home manager? I read [this thread](https://discourse.nixos.org/t/adding-a-flake-package-to-home-manager/21583) and what I got is that I should add the URL to my flake inputs, but after that how would I import the package, maybe putting `noshell.packages.${system}.default` into `home.packages` since I can't make use of it as a `program` not having root access?

**Comment by quazar-omega** [comment]
Solved now by:
1. Adding the flake in `inputs`
2. Adding its provided default package in `home.packages`
3. Configuring the shell file like the README says

(Actual code sample on [my repository](https://codeberg.org/quazar-omega/home-manager-config/src/commit/53558399321f7e169ee5bc293b3b63e1533a5802/flake.nix#L27-L36))
